src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java

```diff
@@ -136,6 +145,47 @@ public static void loadNative() {
     private static final Map<String,PKCS11> moduleMap =
         new HashMap<String,PKCS11>();
 
+    static boolean isKey(CK_ATTRIBUTE[] attrs) {
+        for (CK_ATTRIBUTE attr : attrs) {
+            if ((attr.type == CKA_CLASS) && (attr.getLong() == CKO_SECRET_KEY)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // This is the SunPKCS11 provider instance
+    // there can only be a single PKCS11 provider in
+    // FIPS mode.
+    private static SunPKCS11 mysunpkcs11;
+
+    private static final class InnerPKCS11 extends PKCS11 {
+        InnerPKCS11(String pkcs11ModulePath, String functionListName) throws IOException {
+            super(pkcs11ModulePath, functionListName);
+        }
+
+        // Set PKCS11 instance to FIPS mode, called by SunPKCS11 provider.
+        static void setFIPS(SunPKCS11 sunpkcs11) {
+            mysunpkcs11 = sunpkcs11;
+        }
+
+        // Overriding the JNI method C_CreateObject so that first check if FIPS mode is on and the object is a
+        // secret key, in which case invoke the importKey method in SunPKCS11 provider to import the secret key
+        // into the PKCS11 device.
+        public synchronized long C_CreateObject(long hSession, CK_ATTRIBUTE[] pTemplate) throws PKCS11Exception {
+            if ((mysunpkcs11 != null) && isKey(pTemplate)) {
+                try {
+                    Method method = mysunpkcs11.getClass().getDeclaredMethod("importKey", long.class, CK_ATTRIBUTE[].class);
+                    method.setAccessible(true);
+                    return (Long)method.invoke(mysunpkcs11, hSession, pTemplate);
+                } catch(InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
```

Comment:
space before `(` (here and in `C_CreateObject()`)

Reply:
updated